virtcontainers/utils/utils.go

```diff
@@ -67,6 +68,19 @@ func ReverseString(s string) string {
 	return string(r)
 }
 
+// From the given path to the storage determine
+// if it's a kubernetes ephemeral storage
+func IsEphemeralStroage(path string) bool {
```

Comment:
The check is too k8s specific to be in the virtcontainers library. Please move it to kata cli and use a new device type to describe the ephemeral storage in container config.

Reply:
@bergwolf Need some clarification so just to make sure we are on the same page. 

When you say use a new DeviceType, do you mean the one declared here - https://github.com/kata-containers/runtime/blob/master/virtcontainers/device/config/config.go#L19 ?

Also, you suggested to use container config to describe the storage. Staying within the cli package I could imagine the new device can be added at this point, https://github.com/kata-containers/runtime/blob/master/cli/create.go#L261 (by appending to contConfig.DeviceInfos). Because after that point the code jumps into virtcontainer library.

Coming back to the DeviceType, it's used in the driver implementions in virtcontainer library https://github.com/kata-containers/runtime/tree/master/virtcontainers/device/drivers 

All the uses of the DeviceType including it's declaration is in virtcontainers only. So this isn't helping us taking the code away from virtcontainers and putting it in cli package. 

To me this seems like a overkill to implement all those methods in those drivers when we aren't even going to use any real device from the host. 

Maybe I am missing something, would you like to elaborate more on what you have in mind here?

Reply:
Yes, your understand is correct. I asked for it because this is the device/storage model of virtcontainers. We want all devices/storage to follow the same flow. ephemeral is just another storage. I don't think we should make it a special case and handle it differently.

Besides, `IsEphemeralStroage` is totally bound to k8s. You cannot create an ephemeral storage unless the source path has strict k8s empty dir pattern. That is not flexible at all. We want a generic API that one can create ephemeral storage with any source path. In fact, for ephemeral storage, the source path should not matter at all.

Reply:
> The current code in this PR which uses pb.Storage.Mountpoint is implemented as per our discussion here, #61 (comment)

That comment only describes the fact that the tmpfs storage is already supported by the agent grpc protocol. You have done a great job to implement it in kata agent. But the runtime PR does not satisfy the virtcontainer API we would like to merge.

At the runtime API level, we want an explicit API to let callers specify a tmpfs storage and its mountpoint. It can be a better workaround to check `/proc/mounts` for tmpfs storage, but again, the detection needs to be done at the kata cli level and the virtcontainers API need to be explicit about using a tmpfs storage, instead of trying to do the auto-detection internally.

IOW, what you did is an API change and we need to be explicit about using a tmpfs storage.

Reply:
@harche 

>A typical characteristic of an ephemeral storage would be,
>1. A directory on the host (may or may not be backed by tmpfs)
>2. This directory is bind mounted to all the containers of the same pod so that they can share the data.

I wonder if this is a trusted detection point, I think ephemeral storage has another difference with normal tmpfs storage that a normal tmpfs storage **DOES** allow you to share contents between host and guest but ephemeral storage *DOESN'T* .

So if user wants to share host contents with all containers in pod and this volume happens to base on tmpfs(I think this is also valid use case), the case could be broken.

Reply:
Trying to catch up here, and trying to summarize, we don't have a reliable way to detect this specific use case, right ?
The reason being: a volume backed by `tmpfs` could be either a directory with some data the host want to share with the container, or an ephemeral volume that is an empty directory.

What about checking both the tmpfs and the content of the source path provided ?

Now that being said, I realize that ephemeral volume support in Kata Containers is more an improvement thing since it's already supported if two containers link to the same ephemeral volume on the host via 9p, right @harche ?
Following this, do we really need this "improvement" since it seems to me that we'll have to rely on strong assumptions to detect the case ?

Reply:
We use ephemeral volumes regularly for sharing the data between containers of the pod (especially between init and regular containers). Some of the applications are sensitive to the security and this is where we would love to use kata in our kubernetes environment for the added isolation it provides. With upcoming techonologies like Intel's [TME and MKTME](https://software.intel.com/en-us/blogs/2017/12/22/intel-releases-new-technology-specification-for-memory-encryption), [IBM Ultravisor](https://www.ibm.com/developerworks/library/l-support-protected-computing/) or [AMD memory encryption](http://amd-dev.wpengine.netdna-cdn.com/wordpress/media/2013/12/AMD_Memory_Encryption_Whitepaper_v7-Public.pdf) we can take the isolation provided by kata to the next level. 

Using those technologies you will be able to encrypt the memory pages of the guest, this is where this patch will enable kata to play a crucial role in providing the strongest isolation any runtime can provide for security sensitive kubernetes workloads. Since you will place the ephemeral volume inside the guest, which in turn has encrypted memory pages, even a compromised host won't be able to read the data containers are sharing between them. 

I strongly believe this work aligns with the kata's mission to provide the more secure environment for the container workloads.

@WeiZhang555 @bergwolf @sboeuf @amshinde

Reply:
Thanks for the explanation @harche. This makes sense to support this case, but it's annoying that is something we have to detect based on the path name (as mentioned by @WeiZhang555)